internal/lifecycle/readme.go

```diff
@@ -0,0 +1,156 @@
+// limitations under the License.
+
+package lifecycle
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/GoogleCloudPlatform/serverless-sample-tester/internal/util"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+var (
+	// The tag that should appear immediately before code blocks in a README to indicate that the enclosed commands
+	// are to be used by this program for building and deploying the sample.
+	codeTag = "{sst-run-unix}"
+
+	gcloudCommandRegexp   = regexp.MustCompile(`\bgcloud\b`)
+	cloudRunCommandRegexp = regexp.MustCompile(`\brun\b`)
+
+	gcrURLRegexp = regexp.MustCompile(`gcr.io/.+/\S+`)
+)
+
+// parseREADME parses a README file with the given name. It reads terminal commands surrounded by one of the codeTags
+// listed above and loads them into a Lifecycle. In the process, it replaces the Cloud Run service name and Container
+// Registry tag with the provided inputs.
+func parseREADME(filename, serviceName, gcrURL string) (Lifecycle, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] os.Open %s: %w", filename, err)
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	var lifecycle Lifecycle
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.Contains(line, codeTag) {
+			if s := scanner.Scan(); !s {
+				if err := scanner.Err(); err != nil && !s {
+					return nil, fmt.Errorf("[lifecycle.parseREADME] README bufio.Scanner: %w", err)
+				}
+				return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: unexpected EOF; file ended" +
+					"immediately after code tag")
+			}
+
+			startCodeBlockLine := scanner.Text()
+			c := strings.Contains(startCodeBlockLine, "```")
+			if !c {
+				return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: expecting start of code block" +
+					"immediately after code tag")
+			}
+
+			var blockClosed bool
+			for scanner.Scan() {
+				line = scanner.Text()
+				if strings.Contains(line, "```") {
+					blockClosed = true
+					break
+				}
+
+				line = strings.TrimSpace(line)
+
+				// If there is a backslash at the end of the line, this is a multiline command. Keep scanning to get
+				// entire command.
+				for line[len(line)-1] == '\\' {
+					line = line[:len(line)-1]
+
+					if s := scanner.Scan(); !s {
+						if err := scanner.Err(); err != nil && !s {
+							return nil, fmt.Errorf("[lifecycle.parseREADME] README bufio.Scanner: %w", err)
+						}
+						return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: unexpected EOF; file" +
+							"ended immediately after code tag")
+					}
+
+					l := scanner.Text()
+					if strings.Contains(l, "```") {
+						return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: unexpected end of" +
+							"code block; expecting command line continuation")
+					}
+
+					line = line + strings.TrimSpace(l)
+				}
+
+				line = os.ExpandEnv(line)
+				line = gcrURLRegexp.ReplaceAllString(line, gcrURL)
```

Comment:
Is it possible to separate out the logic involved in just reading the readme from this text substitution? It would involve an additional pass over the data but I think it would make the code clearer (and it doesn't seem like efficiency is the biggest concern here). In this case you could do something like:

* Scan the file and generate a slice of strings that correspond to the code blocks.
* Then write a separate function that turns a code block string into an exec.Cmd by doing substitutions, dealing with multi-line commands, etc

You could even have a separate data type for a code block (`type codeBlock string`) and then give it a method that returns the command.

Reply:
Also I think this could have additional benefits as far as making testing easier, and allowing for support of other input formats other than README should that become necessary.

Reply:
Resolved, thank you.